File: plugin/src/api/export/source.ts
import { AuthorIndex } from "..";
import { Util } from "../../util";
import { Tag } from "./annotations";

export class Source {
	library: number;
	id: number;
	key: string;
	kind: string;
	title: string;
	note?: string;
	date: string;
	url?: string;
	authors: AuthorIndex[];
	tags: Tag[];
	date_added: string;
	date_modified: string;
	path: string;

	constructor(
		library: number,
		id: number,
		key: string,
		kind: string,
		title: string,
		note: string | null,
		date: string,
		url: string | null,
		authors: AuthorIndex[],
		tags: Tag[],
		date_added: string,
		date_modified: string,
		path: string
	) {
		this.library = library;
		this.id = id;
		this.key = key;
		this.kind = kind;
		this.title = title;
		if (note !== null) { this.note = note; };
		this.date = date;
		if (url !== null) { this.url = url; };
		this.authors = authors;
		this.tags = tags;
		this.date_added = date_added;
		this.date_modified = date_modified;
		this.path = path;
	}

	static try_from(item: Zotero.Item): Source | null {
		if (!item.isPDFAttachment()) { return null; }

		let path = item.getFilePath();
		if (path == false) { return null; }

		let parent = item.parentItem;
		if (parent == undefined) { return null; }

		let title: string = parent.getField("shortTitle");
		if (title === "") { title = parent.getDisplayTitle(); }

		// TODO: Check all fields for non-emptiness.

		return new Source(
			parent.libraryID,
			parent.id,
			parent.key,
			parent.itemType,
			title,
			parent.getField("abstractNote"),
			parent.getField("date"),
			parent.getField("url"),
			parent.getCreatorsJSON().filter((c) => c.creatorType === "author").map(AuthorIndex.from),
			parent.getTags().map(Tag.from),
			Zotero.Date.sqlToISO8601(parent.dateAdded),
			Zotero.Date.sqlToISO8601(parent.dateModified),
			path
		);
	}
}

File: importer/src/api/import/source.rs
use std::io;

use chrono::format::StrftimeItems;
use serde::Deserialize;

use crate::{api::shared::{Author, Name}, util::human_date::{self, HumanDate}};

// ! TODO: Sanity check this (eg for non emptiness) on import.
#[allow(unused)]
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SourceImport {
	pub library: i32,
	pub id: i32,
	pub key: String,
	pub kind: String,
	pub title: String,
	pub note: Option<String>,
	pub date: String,
	pub url: Option<String>,
	pub authors: Vec<Author>,
	pub tags: Vec<Tag>,
	pub date_added: String,
	pub date_modified: String,
	pub path: String
}

#[allow(unused)]
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Tag {
	pub name: String
}

impl SourceImport {
	// TODO: Handle errors.
	pub fn parse_date(&self) -> Result<HumanDate, ()> {		
		Ok(HumanDate::parse(&self.date).map_err(|_| ())?)
	}

	pub fn year(&self) -> u32 {
		self.parse_date().unwrap().year
	}

	pub fn primary_author(&self) -> &Author {
		self.authors.iter().next().unwrap()
	}

	pub fn short_name(&self) -> String {
		match &self.primary_author().name {
			Name::Full(full) => format!("{} {}", full.last, self.year()),
			Name::Combined(combined) => format!("{} {}", combined, self.year()),
		}
	}

	pub fn file_name(&self) -> String {
		self.title.clone()
	}
}

#[derive(Debug)]
pub enum ImportSourceError {
	WrongItemCount,
	InvalidItemFormat,
	InvalidJson(serde_path_to_error::Error<serde_json::Error>),
	Filesystem(io::Error)
}

impl From<serde_path_to_error::Error<serde_json::Error>> for ImportSourceError {
	fn from(value: serde_path_to_error::Error<serde_json::Error>) -> Self {
		Self::InvalidJson(value)
	}
}

impl From<io::Error> for ImportSourceError {
	fn from(value: io::Error) -> Self {
		Self::Filesystem(value)
	}
}